### PR TITLE
 XWIKI-19928: Migrate CKEditor to XWiki Platform

### PR DESCRIPTION
**Pull request by manuelleduc**
Jira: https://jira.xwiki.org/browse/XWIKI-19928

* CKEditor history preserved. See https://github.com/xwiki/xwiki-platform/pull/1875/commits/48fceaa6b12fb69ae2e66d31ad584635c23aacdc for my actual changes. I first merged ckeditor at the root (ignoring the conflicting files, then moved the directories to `xwiki-platform-ckeditor` and changed what needed to be adapted).

**Other tasks**
* Move the issues in Jira
* Update the documentation

### CODE REVIEW

**Comment by michitux** [comment]
Commit 6b1053164386aefc526df7512bc664918aa6849b should also be included in this pull request (probably best to check before the merge if there were any new commits).

**Comment by manuelleduc** [comment]
Note: A xar:format call on xwiki-platform-ckeditor-ui needs to be executed.

**Comment by manuelleduc** [comment]
> Note: A xar:format call on xwiki-platform-ckeditor-ui needs to be executed.

done

File: pom.xml [hunk 1]
```diff
@@ -1276,6 +1275,26 @@
...
+            </goals>
+            <configuration>
+              <rules>
+                <bannedDependencies>
+                  <searchTransitive>true</searchTransitive>
+                  <message>
+                    Do not use the CKEditor extension from contrib. Is it now packaged in xwiki platform.
+                  </message>
+                  <excludes>
+                    <exclude>org.xwiki.contrib:application-ckeditor*:*</exclude>
```
Comment (tmortagne) [inline pom.xml:1292]: Why the `:*` ?

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/pom.xml [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+      <groupId>org.xwiki.platform</groupId>
+      <artifactId>xwiki-platform-tree-webjar</artifactId>
+      <version>${project.version}</version>
+      <scope>runtime</scope>
+    </dependency>
+    <!-- This is used by the resource suggest picker on the link modal. -->
+    <dependency>
+      <groupId>org.webjars.npm</groupId>
+      <artifactId>bootstrap-3-typeahead</artifactId>
+      <version>4.0.2</version>
```
Comment (tmortagne) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/pom.xml:67]: Maybe we should define the version of that dependency along with the other bootstrap related dependencies in the root platform pom.

Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/pom.xml:67]: Do we define a variable for every external dependency, even if there is only a single occurrence?

File: pom.xml (hunk 1 above)
Comment (tmortagne) [reply pom.xml:1292]: My point is more that it does not seem needed, `org.xwiki.contrib:application-ckeditor*` should be enough, no ? We generally don't indicate the type there.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/pom.xml (hunk 2 above)
Comment (tmortagne) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/pom.xml:67]: Just suggesting to do the same as https://github.com/xwiki/xwiki-platform/blob/master/pom.xml#L427.

**Pull request merged**